extensions/ql-vscode/test/vscode-tests/test-config.ts

```diff
@@ -1,126 +1,225 @@
+import {
+  ConfigurationScope,
+  ConfigurationTarget,
+  workspace,
+  WorkspaceConfiguration as VSCodeWorkspaceConfiguration,
+} from "vscode";
 import { readFileSync } from "fs-extra";
 import { join } from "path";
-import { ConfigurationTarget } from "vscode";
-import { ALL_SETTINGS, InspectionResult, Setting } from "../../src/config";
 
-class TestSetting<T> {
-  private initialSettingState: InspectionResult<T> | undefined;
+function getIn(object: any, path: string): any {
+  const parts = path.split(".");
+  let current = object;
+  for (const part of parts) {
+    current = current[part];
+    if (current === undefined) {
+      return undefined;
+    }
+  }
+  return current;
+}
+
+function setIn(object: any, path: string, value: any): void {
+  const parts = path.split(".");
+  let current = object;
+  for (const part of parts.slice(0, -1)) {
+    if (current[part] === undefined) {
+      current[part] = {};
+    }
+    current = current[part];
+  }
+  current[parts[parts.length - 1]] = value;
+}
+
+interface WorkspaceConfiguration {
+  scope?: ConfigurationTarget;
+
+  get<T>(section: string | undefined, key: string): T | undefined;
+  has(section: string | undefined, key: string): boolean;
+  update(section: string | undefined, key: string, value: unknown): void;
+}
 
-  constructor(
-    public readonly setting: Setting,
-    private initialTestValue: T | undefined = undefined,
-  ) {}
+class InMemoryConfiguration implements WorkspaceConfiguration {
+  private readonly values: Record<string, unknown> = {};
 
-  public async get(): Promise<T | undefined> {
-    return this.setting.getValue();
+  public constructor(public readonly scope: ConfigurationTarget) {}
+
+  public get<T>(section: string | undefined, key: string): T | undefined {
+    return getIn(this.values, this.getKey(section, key)) as T | undefined;
```

Comment:
Instead of using `getIn` and `setIn`, would it be simpler to just have a simple dictionary that has keys being the entire configuration key? So instead of this:

```
 {
  codeql: {
    cli: {
      executablePath: '/foo/bar/codeql'
    }
  }
}
```

You just have:

```
{
  'codeQL.cli.executablePath': '/foo/bar/codeql'
}
```

What do we get by slicing up the object?

Reply:
The VSCode API also accepts calls like these:

```ts
workspace.getConfiguration("codeQL").update("cli", {
  "executablePath": "/foo/bar/codeql",
}, ConfigurationTarget.Global);
```

If we were storing everything by string key, we'd need to create the correct string keys and remove any string keys which are not passed, but do exist in the configuration.

We're not using any of these calls in the extension, but I think it still makes sense to support as much of the VSCode API as possible if it's not too much work. I think the `getIn` and `setIn` implementations are easier than an implementation by string keys which supports this.